src/main/kotlin/RBSearchTree.kt

```diff
@@ -0,0 +1,276 @@
+                else { // If the node and its parent are not on the same side, then we move them to the same side using rotation
+                    if ( node == node.parent?.leftChild ) {
+                        node = node.parent
+                        rightRotate(node!!)
+                    }// then execute tree balancing
+                    node.parent?.isRed = false
+                    node.parent?.parent?.isRed = true
+                    leftRotate(node.parent?.parent!!)
+                }
+            } else { // execute the same algorithm, but for the case when the uncle is the left child for the grandfather
+                uncle  = node.parent?.parent?.rightChild
+                if ( uncle?.isRed != null && uncle.isRed ) {
+                    uncle.isRed = false
+                    node.parent?.isRed = false
+                    node.parent?.parent?.isRed = true
+                    node = node.parent?.parent
+                } else {
+                    if ( node == node.parent?.rightChild ) {
+                        node = node.parent
+                        leftRotate(node!!)
+                    }
+                    node.parent?.isRed = false
+                    node.parent?.parent?.isRed = true
+                    rightRotate( node.parent?.parent!!)
+                }
+            }
+            if (node == root) { // break the while loop, if the root is reached
+                break
+            }
+        }
+        root?.isRed = false // the root is always black
+    }
+
+    /**
+     * The method returns the root value after inserting the node wiht following key and value parameters
+     * @param K comparable type: key of the node to be inserted
+     * @param V type: value of the node to be inserted
+     */
+    override fun insert(key: K, value: V) {
+        root = insert(root, key, value)
+        root?.isRed = false
+    }
+
+    /**
+     * Searching the insertion place and considering the insertion of the root/children of the root
+     * Return the value of the root after inserting
+     */
+    private fun insert(treeRoot: RedBlackTreeNode<K, V>?, key: K, value: V): RedBlackTreeNode<K, V>? {
+        val node = RedBlackTreeNode(key, value)
+        node.isRed = true
+        var x: RedBlackTreeNode<K, V>? = treeRoot
+        var y: RedBlackTreeNode<K, V>? = null
+        while (x != null) { // searching the place to insert the node
+            y = x
+            x = if (key < x.key) {
+                x.leftChild
+            } else if (key > x.key) {
+                x.rightChild
+            } else if ( x.key == key && x.value == value) { // case of the existence of the node in the tree
+                throw IllegalArgumentException("The key: $key and value: $value already exists in the tree.")
+            } else {
+                throw IllegalArgumentException("The key: $key already exists in the tree.")
+            }
+        }

```

Comment:
Интересное решение не позволять перезаписывать значение по ключу, но в данном случае эти ветки можно объединить так:
```suggestion
            else {
                throw IllegalArgumentException("The key: $key already exists in the tree.")
            }
```